File: core/src/main/java/com/krake/core/media/MediaLoader.java
package com.krake.core.media;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.text.TextUtils;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestBuilder;
import com.bumptech.glide.RequestManager;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.Target;
import com.bumptech.glide.request.transition.DrawableCrossFadeFactory;
import com.krake.core.R;
import com.krake.core.model.MediaPart;

import java.io.UnsupportedEncodingException;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLEncoder;
import java.util.Locale;

/**
 * Questa classe compie tre azione principali:
 * <ul>
 * <li>Crea la richiesta per il WS per il download dell'immagine</li>
 * <li>Scarica l'immagine</li>
 * <li>Setta l'immagine in un widget che implementa l'interfaccia {@link MediaLoadable}, se presente</li>
 * </ul>
 */
@Deprecated
public class MediaLoader {
    private static final String TAG = MediaLoader.class.getSimpleName();

    /**
     * Setta i parametri della richiesta di Glide che verranno passati al WS e verranno utilizzati dal widget corrente
     *
     * @param activity      activity corrente
     * @param mediaLoadable istanza di MediaLoadable
     * @param mediaPart     {@link MediaPart} da caricare
     */
    public static void with(@NonNull Activity activity, @NonNull MediaLoadable mediaLoadable, @Nullable MediaPart mediaPart) {
        with(Glide.with(activity), activity, mediaLoadable, mediaPart);
    }

    /**
     * Setta i parametri della richiesta di Glide che verranno passati al WS e verranno utilizzati dal widget corrente
     *
     * @param fragment      fragment corrente
     * @param mediaLoadable istanza di MediaLoadable
     * @param mediaPart     {@link MediaPart} da caricare
     */
    public static void with(@NonNull Fragment fragment, @NonNull MediaLoadable mediaLoadable, @Nullable MediaPart mediaPart) {
        with(Glide.with(fragment), fragment.getActivity(), mediaLoadable, mediaPart);
    }

    /**
     * Setta i parametri della richiesta di Glide che verranno passati al WS e verranno utilizzati dal widget corrente
     *
     * @param context       context corrente
     * @param mediaLoadable istanza di MediaLoadable
     * @param mediaPart     {@link MediaPart} da caricare
     */
    public static void with(@NonNull Context context, @NonNull MediaLoadable mediaLoadable, @Nullable MediaPart mediaPart) {
        with(Glide.with(context), context, mediaLoadable, mediaPart);
    }

    /**
     * Setta i parametri della richiesta di Glide che verranno passati al WS e verranno utilizzati dal widget corrente
     *
     * @param manager       manager di Glide
     * @param context       context corrente
     * @param mediaLoadable istanza di MediaLoadable
     * @param mediaPart     {@link MediaPart} da caricare
     */
    private static void with(@NonNull RequestManager manager, @NonNull Context context, @NonNull MediaLoadable mediaLoadable, @Nullable MediaPart mediaPart) {
//        if (mediaLoadable instanceof ImageView) {
//            ImageView imageView = (ImageView) mediaLoadable;
            // crea la richiesta

            com.krake.core.media.loader.MediaLoader loader = com.krake.core.media.loader.MediaLoader.Companion.with(context, mediaLoadable);

            if (mediaPart != null)
                loader.mediaPart(mediaPart);

            loader.load();



//            RequestBuilder<Drawable> request = requestWith(manager, context, mediaLoadable, mediaPart);
//            if (request != null) {
//                // setta il target
//                request.into(imageView);
//            }
//        } else {
//            Log.e(TAG, "with: your widget cannot be casted to " + ImageView.class.getCanonicalName());
//        }
    }

    /**
     * Crea la richiesta per Glide nel caso in cui l'url sia valido, in caso contrario, setta il placeholder
     *
     * @param manager       manager di Glide
     * @param context       context corrente
     * @param mediaLoadable istanza di MediaLoadable
     * @param mediaPart     {@link MediaPart} da caricare
     * @return richiesta di Glide
     */
    @SuppressLint("SwitchIntDef")
    public static RequestBuilder<Drawable> requestWith(@NonNull RequestManager manager, @NonNull Context context, @NonNull MediaLoadable mediaLoadable, @Nullable MediaPart mediaPart) {
        ImageOptions options = mediaLoadable.getOptions();
        boolean animated = mediaLoadable.isAnimated();

        int placeholderIdentifier = 0;

        final ImageView imageView = mediaLoadable instanceof ImageView ? (ImageView) mediaLoadable : null;

        if (mediaPart != null) {
            if (!TextUtils.isEmpty(mediaPart.getMediaUrl()) && mediaPart.getMediaType() == MediaType.IMAGE) {
                String mediaUrl = mediaPart.getMediaUrl();

                int glidePlaceholder = mediaLoadable.showPlaceholder() ? mediaLoadable.getPhotoPlaceholder() : 0;

                URL externalURL = null;
                try {
                    externalURL = new URL(mediaUrl);
                } catch (MalformedURLException ignored) {
                }

                RequestBuilder<Drawable> builder;

                if (externalURL == null) {
                    builder = createRequestCreator(manager, context, imageView, options, mediaUrl);
                } else {
                    builder = manager.load(externalURL.toString());
                }

                RequestOptions reqOptions = new RequestOptions();
                if (imageView != null) {
                    // salva lo ScaleType dell'immagine
                    final ImageView.ScaleType imageScaleType = mediaLoadable.getMediaScaleType();
                    final ImageView.ScaleType placeholderScaleType = mediaLoadable.getPlaceholderScaleType();

                    if (mediaLoadable.showPlaceholder() && imageScaleType != placeholderScaleType && imageView.getScaleType() != placeholderScaleType) {
                        imageView.setScaleType(placeholderScaleType);
                    }
                    // se lo ScaleType dell'immagine e quello del placeholder sono differenti bisogna iniziare la procedura per la sostituzione dello ScaleType
                    // settando lo scaleType corretto prima di onResourceReady(), si evita il lag
                    if (imageScaleType == ImageView.ScaleType.FIT_CENTER) {
                        reqOptions.fitCenter();
                    } else if (imageScaleType == ImageView.ScaleType.CENTER_CROP) {
                        reqOptions.centerCrop();
                    }
                    // quando lo ScaleType è differente, probabilmente uno o due frame è possibile che vengano saltati (1 frame ogni 16ms).
                    // questo succede perchè deve richiamare i metodi invalidate() e requestLayout() dello scaleType
                    builder.listener(new RequestListener<Drawable>() {
                        @Override
                        public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                            return false;
                        }

                        @Override
                        public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                            // setta il vecchio scaleType
                            if (imageView.getScaleType() != imageScaleType) {
                                imageView.setScaleType(imageScaleType);
                            }
                            return false;
                        }
                    });
                }

                if (glidePlaceholder > 0) {
                    reqOptions.placeholder(glidePlaceholder);
                }

                if (animated) {
                    DrawableCrossFadeFactory.Builder crossFadeFactoryBuilder = new DrawableCrossFadeFactory.Builder().setCrossFadeEnabled(true);
                    builder.transition(DrawableTransitionOptions.withCrossFade(crossFadeFactoryBuilder));
                } else {
                    reqOptions.dontAnimate();
                }

                reqOptions.diskCacheStrategy(DiskCacheStrategy.DATA);
                builder.apply(reqOptions);
                return builder;
            } else if (mediaLoadable.showPlaceholder()) {
                switch (mediaPart.getMediaType()) {
                    case MediaType.IMAGE:
                        placeholderIdentifier = mediaLoadable.getPhotoPlaceholder();
                        break;
                    case MediaType.VIDEO:
                        placeholderIdentifier = mediaLoadable.getVideoPlaceholder();
                        break;
                    case MediaType.AUDIO:
                        placeholderIdentifier = mediaLoadable.getAudioPlaceholder();
                        break;
                }
            }
        } else if (mediaLoadable.showPlaceholder()) {
            placeholderIdentifier = mediaLoadable.getPhotoPlaceholder();
        }
        if (placeholderIdentifier > 0 && imageView != null) {
            imageView.setScaleType(mediaLoadable.getPlaceholderScaleType());
            imageView.setImageResource(placeholderIdentifier);
        }
        return null;
    }

    /**
     * Crea la richiesta per il WS aggiungendo in querystring tutti parametri stabiliti in app
     *
     * @param manager   manager di Glide
     * @param context   context corrente
     * @param imageView widget di tipo android.widget.ImageView se presente
     * @param options   configurazione da passare al WS
     * @param mediaUrl  url dell'immagine
     * @return richiesta per il WS
     */
    private static RequestBuilder<Drawable> createRequestCreator(@NonNull RequestManager manager, @NonNull Context context, @Nullable ImageView imageView, @NonNull ImageOptions options, @NonNull String mediaUrl) {
        RequestBuilder<Drawable> builder = null;
        int width = options.getWidth();
        int height = options.getHeight();
        boolean sizeIsZero = height == 0 || width == 0;

        if (sizeIsZero && imageView != null) {
            height = imageView.getMeasuredHeight();
            width = imageView.getMeasuredWidth();
        }

        sizeIsZero = height == 0 || width == 0;

        if (sizeIsZero) {
            Resources resources = context.getResources();
            height = resources.getDimensionPixelSize(R.dimen.orchard_image_default_height);
            width = resources.getDimensionPixelSize(R.dimen.orchard_image_default_width);
        }

        String mediaBasePath = Uri.withAppendedPath(Uri.parse(context.getString(R.string.orchard_base_service_url)), context.getString(R.string.orchard_medias_path)).toString();

        try {
            builder = manager.load(String.format(Locale.getDefault(), "%s?Path=%s&Width=%d&Height=%d&Mode=%s&Alignment=%s",
                    mediaBasePath,
                    URLEncoder.encode(mediaUrl, "UTF-8"),
                    width,
                    height,
                    options.getMode(),
                    options.getAlignment()));

        } catch (UnsupportedEncodingException ignored) {
        }
        return builder;
    }

    /**
     * Trasforma un url relativo/parziale in un url assoluto
     *
     * @param context  context corrette
     * @param mediaUrl url del media
     * @return url completo
     */
    public static String getAbsoluteMediaURL(@NonNull Context context, @Nullable String mediaUrl) {
        URL externalURL;
        try {
            externalURL = new URL(mediaUrl);
            return externalURL.toString();
        } catch (MalformedURLException ignored) {
        }
        try {
            return new URL(new URL(context.getString(R.string.orchard_base_service_url)), mediaUrl).toString();
        } catch (MalformedURLException ignored) {
        }
        return null;
    }
}

File: core/src/main/kotlin/com/krake/core/media/loader/MediaLoader.kt
package com.krake.core.media.loader

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.graphics.drawable.Drawable
import android.net.Uri
import android.text.TextUtils
import android.util.Log
import android.widget.ImageView
import androidx.fragment.app.Fragment
import com.krake.core.R
import com.krake.core.media.ImageOptions
import com.krake.core.media.MediaLoadable
import com.krake.core.media.MediaType
import com.krake.core.model.MediaPart
import java.io.UnsupportedEncodingException
import java.net.MalformedURLException
import java.net.URL
import java.net.URLEncoder
import java.util.*

/**
 * Questa classe compie tre azione principali:
 * <ul>
 * <li>Crea la richiesta per il WS per il download dell'immagine</li>
 * <li>Scarica l'immagine</li>
 * <li>Setta l'immagine in un widget che implementa l'interfaccia [MediaLoadable], se presente</li>
 * </ul>
 */
class MediaLoader<T> private constructor(private val context: Context, val mediaLoadable: MediaLoadable) {
    companion object {
        private const val TAG = "MediaLoader"

        fun with(context: Context, mediaLoadable: MediaLoadable) = MediaLoader<Drawable>(context, mediaLoadable)
        fun with(activity: Activity, mediaLoadable: MediaLoadable) = MediaLoader<Drawable>(activity, mediaLoadable)
        fun with(fragment: Fragment, mediaLoadable: MediaLoadable) = MediaLoader<Drawable>(fragment.activity
                ?: throw IllegalArgumentException("The activity mustn't be null"), mediaLoadable)

        fun <T>typedWith(context: Context, mediaLoadable: MediaLoadable): MediaLoader<T> = MediaLoader(context, mediaLoadable)
        fun <T>typedWith(activity: Activity, mediaLoadable: MediaLoadable): MediaLoader<T> = MediaLoader(activity, mediaLoadable)
        fun <T>typedWith(fragment: Fragment, mediaLoadable: MediaLoadable): MediaLoader<T> = MediaLoader(fragment.activity
                ?: throw IllegalArgumentException("The activity mustn't be null"), mediaLoadable)

        fun getAbsoluteMediaURL(context: Context, mediaUrl: String?): String? {
            val externalURL: URL
            try {
                externalURL = URL(mediaUrl)
                return externalURL.toString()
            } catch (ignored: MalformedURLException) {
            }
            try {
                return URL(URL(context.getString(R.string.orchard_base_service_url)), mediaUrl).toString()
            } catch (ignored: MalformedURLException) {
            }
            return null
        }

    }

    private var mediaPart: MediaPart? = null
    private val imageLoader: ImageLoader<String, T> = ImageHandler.loader(context)
    private val listeners = mutableSetOf<ImageLoader.RequestListener<T>>()

    init {
        imageLoader.withCacheStrategy(ImageLoader.CacheStrategy.WITHOUT_TRANSFORMATION)
    }

    fun addListener(vararg listener: ImageLoader.RequestListener<T>) = apply{
        listeners.addAll(listener)
    }

    fun mediaPart(mediaPart: MediaPart) = apply { this.mediaPart = mediaPart }

    fun load() {
        if (mediaLoadable is ImageView) {
            val request = imageLoader.intoView()
            requestWith(imageLoader, context, mediaLoadable, mediaPart, request)?.load(mediaLoadable as ImageView)
        } else {
            Log.e(TAG, "with: your widget cannot be casted to " + ImageView::class.java.canonicalName)
        }
    }

    fun getRequest(): ImageLoader<String, T> {
        requestWith(imageLoader, context, mediaLoadable, mediaPart, null)
        return imageLoader
    }

    /**
     * Crea la richiesta per Glide nel caso in cui l'url sia valido, in caso contrario, setta il placeholder
     *
     * @param manager       manager di Glide
     * @param context       context corrente
     * @param mediaLoadable istanza di MediaLoadable
     * @param mediaPart     [MediaPart] da caricare
     * @return richiesta di Glide
     */
    @SuppressLint("SwitchIntDef")
    private fun requestWith(imageLoader: ImageLoader<String, T>, context: Context, mediaLoadable: MediaLoadable, mediaPart: MediaPart?, request: ImageLoader.ViewRequest?): ImageLoader.ViewRequest? {
        val options = mediaLoadable.options
        val animated = mediaLoadable.isAnimated

        var placeholderIdentifier = 0

        val imageView = if (mediaLoadable is ImageView) mediaLoadable else null

        if (mediaPart != null) {
            if (!TextUtils.isEmpty(mediaPart.mediaUrl) && mediaPart.mediaType == MediaType.IMAGE) {
                val mediaUrl = mediaPart.mediaUrl

                val glidePlaceholder = if (mediaLoadable.showPlaceholder()) mediaLoadable.photoPlaceholder else 0

                var externalURL: URL? = null
                try {
                    externalURL = URL(mediaUrl)
                } catch (ignored: MalformedURLException) {
                }

                if (externalURL == null) {
                    createRequestCreator(imageLoader, context, imageView, options, mediaUrl!!)
                } else {
                    imageLoader.from(externalURL.toString())
                }

                if (imageView != null && request != null) {
                    // salva lo ScaleType dell'immagine
                    val imageScaleType = mediaLoadable.mediaScaleType
                    val placeholderScaleType = mediaLoadable.placeholderScaleType

                    if (mediaLoadable.showPlaceholder() && imageScaleType != placeholderScaleType && imageView.scaleType != placeholderScaleType) {
                        imageView.scaleType = placeholderScaleType
                    }

                    // se lo ScaleType dell'immagine e quello del placeholder sono differenti bisogna iniziare la procedura per la sostituzione dello ScaleType
                    // settando lo scaleType corretto prima di onResourceReady(), si evita il lag
                    request.scaleType(imageScaleType)

                    @Suppress("UNCHECKED_CAST")
                    request.addListener(*listeners.toTypedArray() as Array<out ImageLoader.RequestListener<Drawable>>)
                    // quando lo ScaleType è differente, probabilmente uno o due frame è possibile che vengano saltati (1 frame ogni 16ms).
                    // questo succede perchè deve richiamare i metodi invalidate() e requestLayout() dello scaleType
                    request.addListener(object : ImageLoader.RequestListener<Drawable> {
                        override fun onDataLoadSuccess(resource: Drawable?) {
                            // setta il vecchio scaleType
                            if (imageView.scaleType != imageScaleType) {
                                imageView.scaleType = imageScaleType
                            }
                        }

                        override fun onDataLoadFailed() { }
                    })

                    if (glidePlaceholder > 0) {
                        request.placeHolder(glidePlaceholder)
                    }

                    if (animated) {
                        request.animation(ImageLoader.Animation.FADE)
                    } else {
                        request.animation(ImageLoader.Animation.NONE)
                    }
                }
                return request
            } else if (mediaLoadable.showPlaceholder()) {
                when (mediaPart.mediaType) {
                    MediaType.IMAGE -> placeholderIdentifier = mediaLoadable.photoPlaceholder
                    MediaType.VIDEO -> placeholderIdentifier = mediaLoadable.videoPlaceholder
                    MediaType.AUDIO -> placeholderIdentifier = mediaLoadable.audioPlaceholder
                }
            }
        } else if (mediaLoadable.showPlaceholder()) {
            placeholderIdentifier = mediaLoadable.photoPlaceholder
        }
        if (placeholderIdentifier > 0 && imageView != null) {
            imageView.scaleType = mediaLoadable.placeholderScaleType
            imageView.setImageResource(placeholderIdentifier)
        }
        return null
    }

    /**
     * Crea la richiesta per il WS aggiungendo in querystring tutti parametri stabiliti in app
     *
     * @param manager   manager di Glide
     * @param context   context corrente
     * @param imageView widget di tipo android.widget.ImageView se presente
     * @param options   configurazione da passare al WS
     * @param mediaUrl  url dell'immagine
     * @return richiesta per il WS
     */
    private fun createRequestCreator(imageLoader: ImageLoader<String, T>, context: Context, imageView: ImageView?, options: ImageOptions, mediaUrl: String) {
        var width = options.width
        var height = options.height
        var sizeIsZero = height == 0 || width == 0

        if (sizeIsZero && imageView != null) {
            height = imageView.measuredHeight
            width = imageView.measuredWidth
        }

        sizeIsZero = height == 0 || width == 0

        if (sizeIsZero) {
            val resources = context.resources
            height = resources.getDimensionPixelSize(R.dimen.orchard_image_default_height)
            width = resources.getDimensionPixelSize(R.dimen.orchard_image_default_width)
        }

        val mediaBasePath = Uri.withAppendedPath(Uri.parse(context.getString(R.string.orchard_base_service_url)), context.getString(R.string.orchard_medias_path)).toString()

        try {
            imageLoader.from(String.format(Locale.getDefault(), "%s?Path=%s&Width=%d&Height=%d&Mode=%s&Alignment=%s",
                    mediaBasePath,
                    URLEncoder.encode(mediaUrl, "UTF-8"),
                    width,
                    height,
                    options.mode,
                    options.alignment))

        } catch (ignored: UnsupportedEncodingException) {
        }
    }
}